pkg/compiler/codegen.go

```diff
@@ -1648,15 +1647,131 @@ func (c *codegen) writeJumps(b []byte) error {
 					nextIP-5, offset, math.MaxInt32, math.MinInt32)
 			}
+			if op != opcode.PUSHA && math.MinInt8 <= offset && offset <= math.MaxInt8 {
+				indices = append(indices, ctx.IP())
+			}
+			binary.LittleEndian.PutUint32(arg, uint32(offset))
+		}
+	}
+	// Correct function ip range.
+	// Note: indices are sorted in increasing order.
+	for _, f := range c.funcs {
+	loop:
+		for _, ind := range indices {
+			switch {
+			case ind > int(f.rng.End):
+				break loop
+			case ind < int(f.rng.Start):
+				f.rng.Start -= longToShortRemoveCount
+				f.rng.End -= longToShortRemoveCount
+			case ind >= int(f.rng.Start):
+				f.rng.End -= longToShortRemoveCount
+			}
+		}
+	}
+	return shortenJumps(b, indices), nil
+}
+
+// longToShortRemoveCount is a difference between short and long instruction sizes in bytes.
+const longToShortRemoveCount = 3
+
+// shortenJumps returns converts b to a program where all long JMP*/CALL* specified by indices,
+// are replaced with their corresponding short counterparts. It panics if either b or indices are invalid.
+// This is done in 2 passes:
+// 1. Alter jump offsets taking into account parts to be removed.
+// 2. Perform actual removal of jump targets.
+// Note: after jump offsets altering, there can appear new candidates for conversion.
+// These are ignored for now.
+func shortenJumps(b []byte, indices []int) []byte {
+	if len(indices) == 0 {
+		return b
+	}
+
+	// 1. Alter existing jump offsets.
+	ctx := vm.NewContext(b)
+	for op, _, err := ctx.Next(); err == nil && ctx.IP() < len(b); op, _, err = ctx.Next() {
+		// we can't use arg returned by ctx.Next() because it is copied
+		nextIP := ctx.NextIP()
+		ip := ctx.IP()
+		switch op {
+		case opcode.JMP, opcode.JMPIFNOT, opcode.JMPIF, opcode.CALL,
+			opcode.JMPEQ, opcode.JMPNE,
+			opcode.JMPGT, opcode.JMPGE, opcode.JMPLE, opcode.JMPLT:
+			offset := int(int8(b[nextIP-1]))
+			offset += calcOffsetCorrection(ip, ip+offset, indices)
+			b[nextIP-1] = byte(offset)
+		case opcode.JMPL, opcode.JMPIFL, opcode.JMPIFNOTL,
+			opcode.JMPEQL, opcode.JMPNEL,
+			opcode.JMPGTL, opcode.JMPGEL, opcode.JMPLEL, opcode.JMPLTL,
+			opcode.CALLL, opcode.PUSHA:
+			arg := b[nextIP-4:]
+			offset := int(int32(binary.LittleEndian.Uint32(arg)))
+			offset += calcOffsetCorrection(ip, ip+offset, indices)
 			binary.LittleEndian.PutUint32(arg, uint32(offset))
 		}
 	}
-	return nil
+
+	// 2. Convert instructions.
+	copyOffset := 0
+	l := len(indices)
+	b[indices[0]] = toShortForm(b[indices[0]])
+	for i := 0; i < l; i++ {
+		start := indices[i] + 2
+		end := len(b)
+		if i != l-1 {
+			end = indices[i+1] + 2
+			b[indices[i+1]] = toShortForm(b[indices[i+1]])
+		}
+		copy(b[start-copyOffset:], b[start+3:end])
+		copyOffset += longToShortRemoveCount
+	}
+	return b[:len(b)-copyOffset]
+}
+
+func calcOffsetCorrection(ip, target int, indices []int) int {
+	cnt := 0
+	for _, ind := range indices {
+		if ip <= ind && ind < target ||
+			ind != ip && target <= ind && ind <= ip {
```

Comment:
I think `indices` are naturally ordered, maybe we can use `sort.Search` here to make things a bit more efficient?